lib/interceptor/decompress.js

```diff
@@ -0,0 +1,146 @@
+const { createInflate, createGunzip, createBrotliDecompress } = require('node:zlib')
+const DecoratorHandler = require('../handler/decorator-handler')
+
+class DecompressHandler extends DecoratorHandler {
+  #decompressors = []
+  #skipStatusCodes
+  #skipErrorResponses
+
+  constructor (handler, { skipStatusCodes = [204, 304], skipErrorResponses = true } = {}) {
+    super(handler)
+    this.#skipStatusCodes = skipStatusCodes
+    this.#skipErrorResponses = skipErrorResponses
+  }
+
+  #shouldSkipDecompression (contentEncoding, statusCode) {
+    if (!contentEncoding || statusCode < 200) return true
+    if (this.#skipStatusCodes.includes(statusCode)) return true
+    if (this.#skipErrorResponses && statusCode >= 400) return true
+    return false
+  }
+
+  #createDecompressor (encoding) {
+    const supportedEncodings = {
+      gzip: createGunzip,
+      'x-gzip': createGunzip,
+      deflate: createInflate,
+      br: createBrotliDecompress,
+      compress: createInflate,
+      'x-compress': createInflate
+    }
+
+    const createDecompressor = supportedEncodings[encoding.toLowerCase()]
+    return createDecompressor ? createDecompressor() : null
+  }
+
+  #createDecompressionChain (encodings) {
+    const encodingList = encodings.split(',').map(e => e.trim()).filter(Boolean)
+
+    const decompressors = []
+    for (let i = encodingList.length - 1; i >= 0; i--) {
+      const decompressor = this.#createDecompressor(encodingList[i])
+      if (!decompressor) {
+        return null
+      }
+      decompressors.push(decompressor)
+    }
+
+    return decompressors
+  }
+
+  onResponseStart (controller, statusCode, headers, statusMessage) {
+    const contentEncoding = headers['content-encoding']
+
+    if (this.#shouldSkipDecompression(contentEncoding, statusCode)) {
+      return super.onResponseStart(controller, statusCode, headers, statusMessage)
+    }
+
+    const decompressors = this.#createDecompressionChain(contentEncoding)
+
+    if (!decompressors) {
+      this.#decompressors = []
+      return super.onResponseStart(controller, statusCode, headers, statusMessage)
+    }
+
+    this.#decompressors = decompressors
+
+    const { 'content-encoding': _, 'content-length': __, ...newHeaders } = headers
+
+    if (this.#decompressors.length === 1) {
+      const decompressor = this.#decompressors[0]
+
+      decompressor.on('data', (chunk) => {
+        super.onResponseData(controller, chunk)
+      })
+
+      decompressor.on('error', (error) => {
+        super.onResponseError(controller, error)
+      })
+
+      decompressor.on('finish', () => {
+        super.onResponseEnd(controller, {})
+      })
+    } else {
+      const lastDecompressor = this.#decompressors[this.#decompressors.length - 1]
+
+      for (let i = 0; i < this.#decompressors.length - 1; i++) {
+        this.#decompressors[i].pipe(this.#decompressors[i + 1])
+      }
+
+      lastDecompressor.on('data', (chunk) => {
+        super.onResponseData(controller, chunk)
```

Comment:
We also need to account for backpressure here, in case the reader needs to stop the stream to drain buffer